src/sql/src/plan/typeconv.rs

```diff
@@ -986,15 +989,45 @@ pub fn to_jsonb(ecx: &ExprContext, expr: HirScalarExpr) -> HirScalarExpr {
 /// [union-type-conv]: https://www.postgresql.org/docs/12/typeconv-union-case.html
 pub fn guess_best_common_type(
     ecx: &ExprContext,
-    types: &[Option<ScalarType>],
+    types: &[CoercibleScalarType],
 ) -> Result<ScalarType, PlanError> {
     // This function is a translation of `select_common_type` in PostgreSQL with
     // the addition of our near match logic, which supports Materialize
     // non-linear type promotions.
     // https://github.com/postgres/postgres/blob/d1b307eef/src/backend/parser/parse_coerce.c#L1288-L1308
 
+    // If every type is a literal record with the same number of fields, the
+    // best common type is a record with that number of fields. We recursively
+    // guess the best type for each field.
+    if let Some(CoercibleScalarType::Record(field_tys)) = types.first() {
+        if types
+            .iter()
+            .all(|t| matches!(t, CoercibleScalarType::Record(fts) if field_tys.len() == fts.len()))
+        {
+            let mut fields = vec![];
+            for i in 0..field_tys.len() {
+                let name = ColumnName::from(format!("f{}", fields.len() + 1));
+                let guesses = types
+                    .iter()
+                    .filter_map(|t| match t {
+                        CoercibleScalarType::Record(fts) => {
+                            fts.get(i).map(|ft| ft.clone().scalar_type())
+                        }
+                        _ => unreachable!(),
+                    })
+                    .collect::<Vec<_>>();
+                let guess = guess_best_common_type(ecx, &guesses)?;
+                fields.push((name, guess.nullable(true)));
```

Comment:
The forced nullability here gave me pause for a second––I don't think it's a big deal, but does seem like we could lose potential signal for optimization on `NOT NULL` columns and we do have the information available right here (though it's annoying to assess). I don't have strong feelings because it doesn't change the semantics in any way I can think of but wanted to note.

Reply:
Easy enough to fix. Done!